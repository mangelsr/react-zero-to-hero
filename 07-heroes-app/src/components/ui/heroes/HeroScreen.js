import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router-dom';
import { getHeroesById } from '../../../selectors/getHeroById';


export const HeroScreen = ({ history }) => {

    const { heroId } = useParams();

    const hero = useMemo(() => getHeroesById(heroId), [heroId]);
    
    if (!hero) {
        return <Redirect to="/"/>;
    }

    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    const handleReturn = () => {
        if (history.length <= 2) {
            history.push('/');    
        } else {
            history.goBack();
        }
    };

    return (
        <div className="row">
            <div className="col-4">
                <img className="img-thumbnail animate__animated animate__fadeInLeft" src={`../assets/${heroId}.jpg`}
                alt={superhero} />
            </div>
            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{ superhero }</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">
                        <b>Alter ego:</b> {alter_ego}
                    </li>
                    <li className="list-group-item">
                        <b>Publisher:</b> {publisher}
                    </li>
                    <li className="list-group-item">
                        <b>First Appearance:</b> {first_appearance}
                    </li>
                </ul>
                <h5>Characters</h5>
                <p>{ characters }</p>
                <button onClick={handleReturn} className="btn btn-outline-info">Return</button>
            </div>
        </div>
    );
}
