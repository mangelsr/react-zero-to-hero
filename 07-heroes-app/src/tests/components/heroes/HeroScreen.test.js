import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { HeroScreen } from '../../../components/ui/heroes/HeroScreen';


describe('Tests over <HeroScreen /> component', () => {

    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    };
    
    test('should redirect if no arguments provided', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={historyMock} />
            </MemoryRouter>
        );
        expect(wrapper.find('Redirect').exists()).toBe(true);
    });

    test('should show a hero if param is correct', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={HeroScreen}/>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should return previus screen with push', () => {
        const historyMock = {
            length: 2,
            push: jest.fn(),
            goBack: jest.fn()
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={historyMock} />}/>
            </MemoryRouter>
        );
        wrapper.find('button').simulate('click');
        expect(historyMock.push).toBeCalledTimes(1);
        expect(historyMock.push).toBeCalledWith('/');
        expect(historyMock.goBack).toBeCalledTimes(0);
    });

    test('should return previus screen with goBack', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={historyMock} />}/>
            </MemoryRouter>
        );
        wrapper.find('button').simulate('click');
        expect(historyMock.goBack).toBeCalledTimes(1);
        expect(historyMock.push).not.toBeCalled();
    });


    test('should return redirect if not found', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider-not-found']}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={historyMock} />}/>
            </MemoryRouter>
        );
        expect(wrapper.text()).toBe('');
    });

});