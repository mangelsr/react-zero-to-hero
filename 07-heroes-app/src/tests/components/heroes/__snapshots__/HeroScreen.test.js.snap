// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <HeroScreen /> component should show a hero if param is correct 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego:
        </b>
         
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher:
        </b>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First Appearance:
        </b>
         
        Amazing Fantasy #15
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
